worlds/shapez/regions.py

```diff
@@ -0,0 +1,244 @@
+] + [
+  f"Shapesanity {processing} {coloring}"
+  for processing in shapesanity_processing
+  for coloring in ["Uncolored", "Mixed", "Painted"]
+]
+
+
+def has_cutter(state: CollectionState, player: int) -> bool:
+    return state.has("Cutter", player)
+
+
+def has_rotator(state: CollectionState, player: int) -> bool:
+    # 180 is excluded because of limited functionality
+    return state.has_any(["Rotator", "Rotator (CCW)"], player)
+
+
+def has_stacker(state: CollectionState, player: int) -> bool:
+    return state.has("Stacker", player)
+
+
+def has_painter(state: CollectionState, player: int) -> bool:
+    return state.has_any(["Painter", "Double Painter"], player) or can_use_quad_painter(state, player)
+
+
+def has_mixer(state: CollectionState, player: int) -> bool:
+    return state.has("Color Mixer", player)
+
+
+def has_tunnel(state: CollectionState, player: int) -> bool:
+    return state.has_any(["Tunnel", "Tunnel Tier II"], player)
+
+
+def has_balancer(state: CollectionState, player: int) -> bool:
+    return state.has("Balancer", player) or state.has_all(["Compact Merger", "Compact Splitter"], player)
+
+
+def can_use_quad_painter(state: CollectionState, player: int) -> bool:
+    return state.has_all(["Quad Painter", "Wires"], player) and state.has_any(["Switch", "Constant Signal"], player)
+
+
+def can_build_mam(state: CollectionState, player: int) -> bool:
+    return (has_cutter(state, player) and has_rotator(state, player) and has_stacker(state, player) and
+            has_painter(state, player) and has_mixer(state, player) and has_balancer(state, player) and
+            has_tunnel(state, player) and state.has_all(["Belt Reader", "Storage", "Item Filter", "Wires",
+                                                         "Logic Gates", "Virtual Processing"], player))
+
+
+def can_make_stitched_shape(state: CollectionState, player: int) -> bool:
+    return (state.has_any(["Cutter", "Quad Cutter"], player) and
+            has_rotator(state, player) and has_stacker(state, player))
+
+
+def can_make_east_windmill(state: CollectionState, player: int) -> bool:
+    return (state.has_any(["Cutter", "Quad Cutter"], player) and
+            state.has_any(["Rotator", "Rotator (CCW)", "Rotator (180°)"], player) and
+            has_stacker(state, player))
+
+
+def can_make_half_half_shape(state: CollectionState, player: int) -> bool:
+    return state.has_any(["Cutter", "Quad Cutter"], player) and has_stacker(state, player)
+
+
+def can_make_half_shape(state: CollectionState, player: int) -> bool:
+    return has_cutter(state, player) or state.has_all(["Quad Cutter", "Stacker"], player)
+
+
+def has_logic_list_building(state: CollectionState, player: int, buildings: List[str], index: int,
+                            includeuseful: bool) -> bool:
+    if includeuseful:
+        useful = state.has("Trash", player) and has_balancer(state, player) and has_tunnel(state, player)
+    else:
+        useful = True
```

Comment:
```suggestion
    if includeuseful and not (state.has("Trash", player) and has_balancer(state, player) and has_tunnel(state, player)):
        return False
```
Doing this, you can ditch `useful and ` in the other lines below